src/main/java/core/basesyntax/MyLinkedList.java

```
@@ -3,45 +3,155 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> first;
+    private Node<T> last;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(value, null, null);
+        if (size == 0) {
+            first = newNode;
+            last = newNode;
+        } else {
+            last.next = newNode;
+            newNode.prev = last;
+            last = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Invalid index " + index);
+        }
+        if (index == size) {
+            add(value);
+            return;
+        }
+        Node<T> newNode = new Node<>(value, null, null);
+        if (index == 0) {
+            newNode.next = first;
+            first.prev = newNode;
+            first = newNode;
+        } else {
+            Node<T> current = getNode(index);
+            newNode.prev = current.prev;
+            newNode.next = current;
+            current.prev.next = newNode;
+            current.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (checkIndex(index) == false) {
+            throw new IndexOutOfBoundsException("Element does not exist by index " + index);
```

You can throw an exception in method checkIndex and avoid duplicating code